src/lib/tls/tls12/tls_client_impl_12.cpp

```diff
@@ -754,12 +757,45 @@ void Client_Impl_12::process_handshake_msg(const Handshake_State* active_state,
 
       const bool should_save = save_session(session_info);
 
-      if(!session_id.empty() && state.is_a_resumption() == false)
+      // RFC 5077 3.4
+      //    If the client receives a session ticket from the server, then it
+      //    discards any Session ID that was sent in the ServerHello.
+      const auto handle = [&] -> std::optional<Session_Handle>
          {
-         if(should_save)
-            session_manager().save(session_info);
+         if(!session_ticket.empty())
+            { return Session_Ticket(session_ticket); }
+         else if(!session_id.empty())
+            { return Session_ID(session_id); }
          else
-            session_manager().remove_entry(session_info.session_id());
+            { return std::nullopt; }
+         }();
+
+      if(handle.has_value())
+         {
+         // RFC 5077 3.3
+         //    If the server successfully verifies the client's ticket, then it
+         //    MAY renew the ticket by including a NewSessionTicket handshake
+         //    message after the ServerHello in the abbreviated handshake. The
+         //    client should start using the new ticket as soon as possible
+         //    after it verifies the server's Finished message for new
+         //    connections.
+         if(state.is_a_resumption() &&
+            !state.client_hello()->session_ticket().empty() &&
+            handle->is_ticket())
+            {
+            // renew the session ticket by removing the one we used to establish
+            // this connection and replace it with the one we just received
+            session_manager().remove(Session_Ticket(state.client_hello()->session_ticket())); // TODO: proliferate the strong type
+            session_manager().store(session_info, handle.value());
+            }
```

Comment:
This fixes some BoGo tests that were disabled with "needs investigation". Basically, ticket renewal was not implemented up until now.